Add 'bootloadHID' flash target (#5587)
qmk/qmk_firmware: docs/flashing_bootloadhid.md

# BootloadHID Flashing Instructions and Bootloader Information

> ps2avr(GB) boards use an atmega32a microcontroller and a different bootloader. It is not flashable using the regular QMK methods.

Reviewer: ```suggestion
ps2avr(GB) boards use an ATmega32A microcontroller and a different bootloader. It is not flashable using the regular QMK methods.
```

General flashing sequence:

1. Enter the bootloader using any of the following methods:
    * Tap the `RESET` keycode (may not work on all devices)
    * Hold the salt key while plugging the keyboard in (usually documented within keyboard readme) 
2. Wait for the OS to detect the device
3. Flash a .hex file
4. Reset the device into application mode (may be done automatically)

## bootloadHID Flashing Target

> Using the qmk installation script, detailed [here](newbs_getting_started.md), the required bootloadHID tools should be automatically installed.

Reviewer: ```suggestion
Using the QMK installation script, detailed [here](newbs_getting_started.md), the required bootloadHID tools should be automatically installed.
```

> To flash via the command line, use the target `:bootloadHID` by execute the following command:

Reviewer: ```suggestion
To flash via the command line, use the target `:bootloadHID` by executing the following command:
```

    make <keyboard>:<keymap>:bootloadHID

## GUI Flashing

### Windows
1. Download [HIDBootFlash](http://vusb.wikidot.com/project:hidbootflash).
2. Place your keyboard into reset. 

> 3. Ensure the configured VendorID is `16c0` and ProductID is `05df`

Reviewer: ```suggestion
3. Ensure the configured USB Vendor ID is `16C0` and Product ID is `05DF`.
```

Author: The terminology used was intended to match the applications usage,
![image](https://user-images.githubusercontent.com/16520359/60147487-78a5e400-97c5-11e9-8ff9-3111e1202070.png). Would bold/italic help in this instance?

Reviewer: Should be easy enough to grok, but I removed the USB part just in case.

4. Press the `Find Device` button and ensure that your keyboard is found.
5. Press the `Open .hex File` button and locate the `.hex` file you created.
6. Press the `Flash Device` button and wait for the process to complete.

## Command Line Flashing

1. Place your keyboard into reset. 
2. Flash the board by typing `bootloadHID -r` followed by the path to your `.hex` file.

### Windows Manual Installation
For MSYS2:

> 1. Download the BootloadHID firmware package from https://www.obdev.at/downloads/vusb/bootloadHID.2012-12-08.tar.gz

Reviewer: ```suggestion
1. Download the BootloadHID firmware package from https://www.obdev.at/downloads/vusb/bootloadHID.2012-12-08.tar.gz.
```

> 2. Extract contents using a compatible tool, for example 7zip

Reviewer: ```suggestion
2. Extract contents using a compatible tool, for example 7-Zip.
```

3. Add to the MSYS path by copying `commandline/bootloadHID.exe` from the extracted archive to your MSYS2 installation, typically `C:\msys64\usr\bin`

For native Windows flashing, the `bootloadHID.exe` can be used outside of the MSYS2 environment.

### Linux Manual Installation
1. Install libusb development dependency:
    ```bash

>     # This depends on OS - for debiant the following works

Reviewer: ```suggestion
    # This depends on OS - for Debian the following works
```

    sudo apt-get install libusb-dev
    ```
2. Download the BootloadHID firmware package:
    ```
    wget https://www.obdev.at/downloads/vusb/bootloadHID.2012-12-08.tar.gz -O - | tar -xz -C /tmp
    ```
3. Build the bootloadHID executable:
    ```
    cd /tmp/bootloadHID.2012-12-08/commandline/
    make
    sudo cp bootloadHID /usr/local/bin
    ```

### MacOS Manual Installation

> 1. Install homebrew by typing the following:

Reviewer: ```suggestion
1. Install Homebrew by typing the following:
```

### Windows Manual Installation
For MSYS2:
1. Download the BootloadHID firmware package from https://www.obdev.at/downloads/vusb/bootloadHID.2012-12-08.tar.gz
2. Extract contents using a compatible tool, for example 7zip
3. Add to the MSYS path by copying `commandline/bootloadHID.exe` from the extracted archive to your MSYS2 installation, typically `C:\msys64\usr\bin`

For native Windows flashing, the `bootloadHID.exe` can be used outside of the MSYS2 environment.

### Linux Manual Installation
1. Install libusb development dependency:
    ```bash
    # This depends on OS - for debiant the following works
    sudo apt-get install libusb-dev
    ```
2. Download the BootloadHID firmware package:
    ```
    wget https://www.obdev.at/downloads/vusb/bootloadHID.2012-12-08.tar.gz -O - | tar -xz -C /tmp
    ```
3. Build the bootloadHID executable:
    ```
    cd /tmp/bootloadHID.2012-12-08/commandline/
    make
    sudo cp bootloadHID /usr/local/bin
    ```

### MacOS Manual Installation
1. Install homebrew by typing the following:
    ```
    /usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
    ```
2. Install `crosspack-avr`:
    ```

>     brew cask install crosspack-avr

Reviewer: Is this cask necessary for flashing with BootloadHID, or can the `osx-cross/homebrew-avr` tap method be substituted here? IMO the latter is preferable since it sources binutils and gcc directly (though it does take some time to compile).

Author: At the moment that is a copy-paste from the old docs. Until i repurpose some old hardware, i dont have access to anything that can test that suggestion so will need some help testing.

Reviewer: Whoops, didn't see it was deleted further down. Carry on.

Author: I want these centralised docs to be rock solid so i'm happy for anysuggestions on how to make them better.

The `crosspack-avr` dependency falls into the same category as the python2/python3 dependency i have also questioned.

Reviewer: Seems like crosspack-avr is just another AVR toolchain, so unless BootloadHID for some reason requires something provided by one, I'm thinking it doesn't need to be mentioned at all.

Author: Finally got a chance to test the install requirements on mac, and none of them are required.

    ```
3. Install the following packages:
    ```

>     brew install python3

Author: The python dependencies are only required for the `atmega32a_program.py`. As this PR aims to remove the need for `:program` with the new `:bootloadHID` target, these dependencies should be removed from the docs?
